app/src/common/shared/org/mozilla/vrbrowser/telemetry/GleanMetricsService.java

```diff
@@ -0,0 +1,56 @@
+package org.mozilla.vrbrowser.telemetry;
+
+import android.content.Context;
+import android.util.Log;
+
+import androidx.annotation.UiThread;
+
+import org.mozilla.vrbrowser.browser.SettingsStore;
+import org.mozilla.vrbrowser.search.SearchEngineWrapper;
+import org.mozilla.vrbrowser.utils.SystemUtils;
+import mozilla.components.service.glean.Glean;
+
+
+public class GleanMetricsService {
+
+    private static boolean initialized = false;
+    private final static String LOGTAG = SystemUtils.createLogtag(GleanMetricsService.class);
+    //private static GleanInternalAPI glean = null;
+    private static Context context = null;
+
+    // We should call this at the application initial stage. Instead,
+    // it would be called when users turn on/off the setting of telemetry.
+    // e.g., SettingsStore.getInstance(context).setTelemetryEnabled();
+    public static void init(Context aContext) {
+        if (initialized)
+            return;
+
+        context = aContext;
+        initialized = true;
+
+        final boolean telemetryEnabled = SettingsStore.getInstance(aContext).isTelemetryEnabled();
+
+        // TODO: Can't find symbol variable pings. Confirm if we need a ping.yaml?
+        //Glean.INSTANCE.registerPings(Pings);
+        Glean.INSTANCE.setUploadEnabled(telemetryEnabled);
+        Glean.INSTANCE.initialize(aContext);
```

Comment:
@Dexterp37 is there any way to let Glean allow us configure as below?

```
final TelemetryConfiguration configuration = new TelemetryConfiguration(aContext)
                    .setServerEndpoint("https://incoming.telemetry.mozilla.org")
                    .setAppName(APP_NAME + "_" + (DeviceType.isOculusBuild() ? "oculusvr" : BuildConfig.FLAVOR_platform))
                    .setUpdateChannel(BuildConfig.BUILD_TYPE)
                    .setPreferencesImportantForTelemetry(resources.getString(R.string.settings_key_locale))
                    .setCollectionEnabled(telemetryEnabled)
                    .setUploadEnabled(telemetryEnabled)
                    .setBuildId(String.valueOf(BuildConfig.VERSION_CODE));
```
I am looking for `AppName` and `BuildId` in `Configuration.kt`, but it seems like it doesn't have.

Reply:
Hi @daoshengmu . The Glean SDK works a bit differently, and infers most of these fields on its own. See the [pings docs](https://mozilla.github.io/glean/book/user/pings/index.html#the-client_info-section).

- The `BuildId` is automatically inferred by the Glean SDK and automatically reported in the `client_info.app_build` field in all pings;
- As for the `AppName`, what is this used for? The Glean SDK automatically reports the `application_id` with all its pings. Do you have a different application id for oculus builds?
- `appChannel` can be passed through the `Configuration` object.
- `serverEndpoint` does not need to be provided. If required by the Java, please use `mozilla.components.service.glean.config.Configuration.DEFAULT_TELEMETRY_ENDPOINT`.

Reply:
@Dexterp37 Our `BuildId` is dynamic generated by the taskcluster builds, and it comes from the versionCode in  `app/build.gradle`. If the `BuildId` is using versionCode from `app/build.gradle` that will be fine. Otherwise, I would suggest giving us this ability to set `BuildId`.

`AppName` is a tricky one. Our project has multiple build configurations for different platforms. Our project will deploy to Oculus, HTC Wave, Google Play app stores, so their app name are different (`FirefoxReality_wavevr`, `FirefoxReality_oculusvr`, `FirefoxReality_googlevr`). I am also wondering if we have a way to set it manually.

Reply:
> 
> 
> @Dexterp37 Our `BuildId` is dynamic generated by the taskcluster builds, and it comes from the versionCode in `app/build.gradle`. If the `BuildId` is using versionCode from `app/build.gradle` that will be fine. Otherwise, I would suggest giving us this ability to set `BuildId`.

Yes, it comes from `versionCode`. Please double check you're getting the expected value once you are done with the integration.

> `AppName` is a tricky one. Our project has multiple build configurations for different platforms. Our project will deploy to Oculus, HTC Wave, Google Play app stores, so their app name are different (`FirefoxReality_wavevr`, `FirefoxReality_oculusvr`, `FirefoxReality_googlevr`). I am also wondering if we have a way to set it manually.

What is the application name used for? What is the question you're trying to answer with this?
Do applications deployed to different stores have different application ids?

Reply:
> > @Dexterp37
> > Our app ids are the same in these three stores.
> > Regarding to `AppName`, there is a field in the original telemetry, `app_name`.[1] It will be the product name, for example, `Fenix`, `Fennec`, etc... For FirefoxReality, our product name is `FirefoxReality`, but in order to get different three product information of Oculus, HTC Wave, Google Play. We added a suffix to help us recognize them in queries. (`FirefoxReality_wavevr`, `FirefoxReality_oculusvr`, `FirefoxReality_googlevr`)
> > [1] https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/data/core-ping.html
> 
> Thank you for the details. I'm sorry but I need to keep asking questions in order to better understand your intent here :)
> 
> We stopped using the application name as an identifier since it allows for easy collision and tampering on the client side. By using the application id, we can guarantee that no application with that id will send data from one _store_.
> 
> @daoshengmu, if the intent is to distinguish the 3 products, may I ask why not use the device manufacturer or model?

@Dexterp37 For Oculus platform, that's ok because they only have Go, and Quest. But for GoogleVR and HTC Wave, both of them are open platforms, lots of manufacturers are very new to me. If you are interested to see, we have a query to list these manufactures to us. BTW, I saw Fenix are still using `app_name`, IIUC, Glean doesn't have this field in the baseline ping, if so, how do they get this data in BigQuery [1]?

https://github.com/mozilla/bigquery-etl/blob/d0ca506/sql/telemetry/firefox_nondesktop_exact_mau28_raw_v1/query.sql#L57-L64.[1]

Reply:
> > @daoshengmu, if the intent is to distinguish the 3 products, may I ask why not use the device manufacturer or model?
> 
> @Dexterp37 For Oculus platform, that's ok because they only have Go, and Quest. But for GoogleVR and HTC Wave, both of them are open platforms, lots of manufacturers are very new to me. If you are interested to see, we have a query to list these manufactures to us.

It seems to me that you're using this to tell different distributions apart, rather than for having application with different names. I think there's a bunch of options here that you to consider:

1. use different application ids for the different stores; this would make data flow into different tables while still giving you the opportunity to join it if needed; this might have implications for users, since renaming application ids might require to reinstall the application or some engineering might be needed to migrate old app ids to new ones;

2. add the distribution suffix to your [`versionName`](https://developer.android.com/studio/publish/versioning.html). This will be reported as the `app_display_version` by Glean. As described in the Android docs, you can specify different application flavours so you can simply have one for each distribution; 

3. add a new metric to the Glean SDK, in the `metrics.yaml`, call it something like `distribution_name` and set it to either `wavevr`, `oculusvr` or `googlevr` at run-time. Note that you would need to change the `send_in_pings` in the registry file every time you add a new custom ping.

I'd slightly prefer (2), but (3) would be a good option as well.

> BTW, I saw Fenix are still using `app_name`, IIUC, Glean doesn't have this field in the baseline ping, if so, how do they get this data in BigQuery [1]?
> 
> [https://github.com/mozilla/bigquery-etl/blob/d0ca506/sql/telemetry/firefox_nondesktop_exact_mau28_raw_v1/query.sql#L57-L64.[1]](https://github.com/mozilla/bigquery-etl/blob/d0ca506/sql/telemetry/firefox_nondesktop_exact_mau28_raw_v1/query.sql#L57-L64.%5B1%5D)

This is an implementation detail of the [clients_daily ETL](https://github.com/mozilla/bigquery-etl/blob/master/templates/org_mozilla_fenix_derived/clients_daily_v1/query.sql#L44). That name does not come from the SDK on the client side. The only things we send are the ones in [our docs](https://mozilla.github.io/glean/book/user/pings/index.html).